# 01 — Data Inventory

## Dengue Disease Prediction for Sri Lanka

#### Purpose

The purpose of this notebook is to inspect the raw datasets used in the study before performing any preprocessing.

The three main data sources are:

1. Dengue case data
2. Weather data
3. Population data

This notebook examines the file structure, dimensions, column names, data types, missing values, duplicates, temporal coverage and geographic coverage of the datasets.

No raw data are modified in this notebook.

In [1]:
import pandas as pd
from pathlib import Path

In [2]:
PROJECT_ROOT = Path.cwd()
if PROJECT_ROOT.name == "notebooks":
    PROJECT_ROOT = PROJECT_ROOT.parent
print(PROJECT_ROOT)

d:\Dengue_Project


In [3]:
DENGUE_DIR = PROJECT_ROOT / "data" / "raw" / "dengue"
WEATHER_DIR = PROJECT_ROOT / "data" / "raw" / "weather"
POPULATION_DIR = PROJECT_ROOT / "data" / "raw" / "population"

print("Dengue folder:", DENGUE_DIR)
print("Weather folder:", WEATHER_DIR)
print("Population folder:", POPULATION_DIR)

Dengue folder: d:\Dengue_Project\data\raw\dengue
Weather folder: d:\Dengue_Project\data\raw\weather
Population folder: d:\Dengue_Project\data\raw\population


In [4]:
print("Dengue folder exists:", DENGUE_DIR.exists())
print("Weather folder exists:", WEATHER_DIR.exists())
print("Population folder exists:", POPULATION_DIR.exists())

Dengue folder exists: True
Weather folder exists: True
Population folder exists: True


In [6]:
dengue_file = list(DENGUE_DIR.glob("*.csv"))[0]
weather_file = list(WEATHER_DIR.glob("*.csv"))[0]
population_file = list(POPULATION_DIR.glob("*.csv"))[0]

print("Dengue:", dengue_file.name)
print("Weather:", weather_file.name)
print("Population:", population_file.name)

Dengue: srilanka_dengue_weekly_2010_2024.csv
Weather: sri_lanka_weekly_weather_2010_2024.csv
Population: district_population_2014_2024.csv


In [7]:
dengue = pd.read_csv(dengue_file)
weather = pd.read_csv(weather_file)
population = pd.read_csv(population_file)

Check the dataset sizes

In [8]:
print("Dengue shape:", dengue.shape)
print("Weather shape:", weather.shape)
print("Population shape:", population.shape)

Dengue shape: (20306, 6)
Weather shape: (20332, 19)
Population shape: (475, 3)


In [9]:
display(dengue.head())

,year,week,start.date,end.date,district,cases
0,2010,1,1/2/2010,1/8/2010,Colombo,176
1,2010,1,1/2/2010,1/8/2010,Gampaha,153
2,2010,1,1/2/2010,1/8/2010,Kalutara,32
3,2010,1,1/2/2010,1/8/2010,Kandy,68
4,2010,1,1/2/2010,1/8/2010,Matale,25


In [10]:
display(weather.head())

,year,week,district,latitude,longitude,period_start,period_end,temperature_mean_c,temperature_min_c,temperature_max_c,humidity_mean_percent,humidity_min_percent,humidity_max_percent,rainfall_total_mm,rainfall_max_hourly_mm,rainy_days,wind_speed_mean_kmh,wind_speed_max_kmh,source_file
0,2010,2,Ampara,7.29,81.67,2010-01-02,2010-01-08,25.06,21.9,28.6,77.10,59,91,5.8,0.6,4,12.47,21.1,Ampara_weekly_weather_2010_2024.csv
1,2010,2,Anuradhapura,8.31,80.40,2010-01-02,2010-01-08,23.99,18.9,28.1,79.37,55,98,34.9,4.4,3,9.45,18.4,Anuradhapura_weekly_weather_2010_2024.csv
2,2010,2,Badulla,6.99,81.06,2010-01-02,2010-01-08,22.27,18.5,26.5,83.08,59,98,9.9,1.6,3,5.30,10.5,Badulla_weekly_weather_2010_2024.csv
3,2010,2,Batticaloa,7.72,81.70,2010-01-02,2010-01-08,25.20,21.9,28.2,78.53,59,96,9.2,0.8,5,11.12,20.0,Batticaloa_weekly_weather_2010_2024.csv
4,2010,2,Colombo,6.93,79.86,2010-01-02,2010-01-08,25.50,21.3,30.1,74.12,52,94,1.1,0.2,2,12.82,19.8,Colombo_weekly_weather_2010_2024.csv


In [11]:
display(population.head())

,district,year,population
0,Colombo,2014,2353000
1,Colombo,2015,2367000
2,Colombo,2016,2387000
3,Colombo,2017,2407000
4,Gampaha,2014,2339000


Check the column names

In [12]:
print("DENGUE COLUMNS")
print(dengue.columns.tolist())
print("\nWEATHER COLUMNS")
print(weather.columns.tolist())
print("\nPOPULATION COLUMNS")
print(population.columns.tolist())

DENGUE COLUMNS
['year', 'week', 'start.date', 'end.date', 'district', 'cases']

WEATHER COLUMNS
['year', 'week', 'district', 'latitude', 'longitude', 'period_start', 'period_end', 'temperature_mean_c', 'temperature_min_c', 'temperature_max_c', 'humidity_mean_percent', 'humidity_min_percent', 'humidity_max_percent', 'rainfall_total_mm', 'rainfall_max_hourly_mm', 'rainy_days', 'wind_speed_mean_kmh', 'wind_speed_max_kmh', 'source_file']

POPULATION COLUMNS
['district', 'year', 'population']


Check data types

In [ ]:
print("DENGUE DATA TYPES")
print(dengue.dtypes)
print("\nWEATHER DATA TYPES")
print(weather.dtypes)
print("\nPOPULATION DATA TYPES")
print(population.dtypes)

DENGUE DATA TYPES
year          int64
week          int64
start.date      str
end.date        str
district        str
cases         int64
dtype: object

WEATHER DATA TYPES
year                        int64
week                        int64
district                      str
latitude                  float64
longitude                 float64
period_start                  str
period_end                    str
temperature_mean_c        float64
temperature_min_c         float64
temperature_max_c         float64
humidity_mean_percent     float64
humidity_min_percent        int64
humidity_max_percent        int64
rainfall_total_mm         float64
rainfall_max_hourly_mm    float64
rainy_days                  int64
wind_speed_mean_kmh       float64
wind_speed_max_kmh        float64
source_file                   str
dtype: object

POPULATION DATA TYPES
district        str
year          int64
population    int64
dtype: object


Check missing values

In [14]:
print("DENGUE MISSING VALUES")
print(dengue.isnull().sum())
print("\nWEATHER MISSING VALUES")
print(weather.isnull().sum())
print("\nPOPULATION MISSING VALUES")
print(population.isnull().sum())

DENGUE MISSING VALUES
year          0
week          0
start.date    0
end.date      0
district      0
cases         0
dtype: int64

WEATHER MISSING VALUES
year                      0
week                      0
district                  0
latitude                  0
longitude                 0
period_start              0
period_end                0
temperature_mean_c        0
temperature_min_c         0
temperature_max_c         0
humidity_mean_percent     0
humidity_min_percent      0
humidity_max_percent      0
rainfall_total_mm         0
rainfall_max_hourly_mm    0
rainy_days                0
wind_speed_mean_kmh       0
wind_speed_max_kmh        0
source_file               0
dtype: int64

POPULATION MISSING VALUES
district      0
year          0
population    0
dtype: int64


Check duplicated rows

In [15]:
print("Dengue duplicate rows:", dengue.duplicated().sum())
print("Weather duplicate rows:", weather.duplicated().sum())
print("Population duplicate rows:", population.duplicated().sum())

Dengue duplicate rows: 0
Weather duplicate rows: 0
Population duplicate rows: 200


We have 200 duplication rows in the population dataset

In [16]:
for column in weather.columns:
    print(column)

year
week
district
latitude
longitude
period_start
period_end
temperature_mean_c
temperature_min_c
temperature_max_c
humidity_mean_percent
humidity_min_percent
humidity_max_percent
rainfall_total_mm
rainfall_max_hourly_mm
rainy_days
wind_speed_mean_kmh
wind_speed_max_kmh
source_file


## Temporal Coverage

This section examines the time periods covered by the dengue, weather and population datasets.

Understanding temporal coverage is necessary before integrating the datasets because the datasets must have compatible time periods for forecasting.

In [19]:
print("Dengue years:", dengue["year"].min(), "-", dengue["year"].max())
print("Weather years:", weather["year"].min(), "-", weather["year"].max())
print("Population years:", population["year"].min(), "-", population["year"].max())

Dengue years: 2010 - 2024
Weather years: 2010 - 2024
Population years: 2014 - 2024


In [20]:
print("Dengue years:")
print(sorted(dengue["year"].unique()))

print("\nWeather years:")
print(sorted(weather["year"].unique()))

print("\nPopulation years:")
print(sorted(population["year"].unique()))

Dengue years:
[np.int64(2010), np.int64(2011), np.int64(2012), np.int64(2013), np.int64(2014), np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020), np.int64(2021), np.int64(2022), np.int64(2023), np.int64(2024)]

Weather years:
[np.int64(2010), np.int64(2011), np.int64(2012), np.int64(2013), np.int64(2014), np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020), np.int64(2021), np.int64(2022), np.int64(2023), np.int64(2024)]

Population years:
[np.int64(2014), np.int64(2015), np.int64(2016), np.int64(2017), np.int64(2018), np.int64(2019), np.int64(2020), np.int64(2021), np.int64(2022), np.int64(2023), np.int64(2024)]


In [21]:
print(dengue["year"].value_counts().sort_index())
print(weather["year"].value_counts().sort_index())
print(population["year"].value_counts().sort_index())

year
2010    1352
2011    1352
2012    1352
2013    1352
2014    1352
2015    1352
2016    1378
2017    1352
2018    1352
2019    1352
2020    1352
2021    1378
2022    1352
2023    1326
2024    1352
Name: count, dtype: int64
year
2010    1352
2011    1352
2012    1352
2013    1352
2014    1352
2015    1352
2016    1378
2017    1352
2018    1352
2019    1352
2020    1352
2021    1378
2022    1352
2023    1352
2024    1352
Name: count, dtype: int64
year
2014    50
2015    50
2016    50
2017    50
2018    50
2019    50
2020    50
2021    50
2022    25
2023    25
2024    25
Name: count, dtype: int64


## Weekly Coverage

The dengue and weather datasets contain weekly observations. This section checks the number of unique weeks available in each year and identifies any differences in weekly coverage between the two datasets.

In [22]:
dengue_weeks = dengue.groupby("year")["week"].nunique()
print(dengue_weeks)

year
2010    52
2011    52
2012    52
2013    52
2014    52
2015    52
2016    53
2017    52
2018    52
2019    52
2020    52
2021    53
2022    52
2023    51
2024    52
Name: week, dtype: int64


In [23]:
weather_weeks = weather.groupby("year")["week"].nunique()
print(weather_weeks)

year
2010    52
2011    52
2012    52
2013    52
2014    52
2015    52
2016    53
2017    52
2018    52
2019    52
2020    52
2021    53
2022    52
2023    52
2024    52
Name: week, dtype: int64


In [24]:
week_comparison = pd.DataFrame({
    "dengue_weeks": dengue_weeks,
    "weather_weeks": weather_weeks
})

week_comparison

,dengue_weeks,weather_weeks
year,,
2010,52,52
2011,52,52
2012,52,52
2013,52,52
2014,52,52
2015,52,52
2016,53,53
2017,52,52
2018,52,52


In [25]:
dengue_2023_weeks = set(dengue[dengue["year"] == 2023]["week"])

weather_2023_weeks = set(weather[weather["year"] == 2023]["week"])

print("Weeks in weather but not dengue:")
print(weather_2023_weeks - dengue_2023_weeks)

print("Weeks in dengue but not weather:")
print(dengue_2023_weeks - weather_2023_weeks)

Weeks in weather but not dengue:
{52}
Weeks in dengue but not weather:
set()


In [26]:
print("Dengue rows for 2023 week 52:")
print(len(dengue[(dengue["year"] == 2023) & (dengue["week"] == 52)]))

print("Weather rows for 2023 week 52:")
print(len(weather[(weather["year"] == 2023) & (weather["week"] == 52)]))

Dengue rows for 2023 week 52:
0
Weather rows for 2023 week 52:
26


### Observation: Weekly Coverage

The dengue and weather datasets have matching weekly coverage for most years. Years 2016 and 2021 contain 53 weeks in both datasets.

A temporal coverage difference was identified in 2023. Week 52 is present in the weather dataset but absent from the dengue dataset. This accounts for the 26 row difference between the weather and dengue datasets, corresponding to the 26 dengue reporting areas.

This issue is recorded for later preprocessing. No values are added or modified in the raw datasets in this notebook.

understand the population duplicates

In [27]:
print("Total population rows:", len(population))
print("Unique population rows:", len(population.drop_duplicates()))
print("Number of districts:", population["district"].nunique())
print("Number of years:", population["year"].nunique())

Total population rows: 475
Unique population rows: 275
Number of districts: 25
Number of years: 11


In [28]:
len(population.drop_duplicates())

275

### Observation: Population Duplicates

The raw population dataset contains 475 rows, of which 275 are unique records. The dataset covers 25 districts across 11 years (2014–2024), corresponding to 275 unique district-year combinations.

A total of 200 exact duplicate rows were identified. These duplicates will be handled during population data preprocessing. No rows are removed in this data inventory notebook.

## Geographic Coverage

This section examines the geographic units represented in the dengue, weather and population datasets.

Geographic compatibility is important because district names must be standardized before the datasets can be integrated.

In [29]:
print("Dengue areas:", dengue["district"].nunique())
print("Weather areas:", weather["district"].nunique())
print("Population areas:", population["district"].nunique())

Dengue areas: 26
Weather areas: 26
Population areas: 25


In [30]:
print("DENGUE AREAS")
for name in sorted(dengue["district"].unique()):
    print(name)

DENGUE AREAS
Ampara
Anuradhapura
Badulla
Batticaloa
Colombo
Galle
Gampaha
Hambanthota
Jaffna
Kalmune
Kalutara
Kandy
Kegalle
Kilinochchi
Kurunegala
Mannar
Matale
Matara
Monaragala
Mullaitivu
NuwaraEliya
Polonnaruwa
Puttalam
Ratnapura
Trincomalee
Vavuniya


In [31]:
print("WEATHER AREAS")
for name in sorted(weather["district"].unique()):
    print(name)

WEATHER AREAS
Ampara
Anuradhapura
Badulla
Batticaloa
Colombo
Galle
Gampaha
Hambantota
Jaffna
Kalmunai
Kalutara
Kandy
Kegalle
Kilinochchi
Kurunegala
Mannar
Matale
Matara
Monaragala
Mullaitivu
Nuwara Eliya
Polonnaruwa
Puttalam
Ratnapura
Trincomalee
Vavuniya


In [32]:
print("POPULATION AREAS")
for name in sorted(population["district"].unique()):
    print(name)

POPULATION AREAS
Ampara
Anuradhapura
Badulla
Batticaloa
Colombo
Galle
Gampaha
Hambantota
Jaffna
Kalutara
Kandy
Kegalle
Kilinochchi
Kurunegala
Mannar
Matale
Matara
Monaragala
Mullaitivu
Nuwara-eliya
Polonnaruwa
Puttalam
Ratnapura
Trincomalee
Vavuniya


In [33]:
dengue_areas = set(dengue["district"].unique())
weather_areas = set(weather["district"].unique())
population_areas = set(population["district"].unique())

In [34]:
print("In dengue but NOT in weather:")
print(dengue_areas - weather_areas)

print("\nIn weather but NOT in dengue:")
print(weather_areas - dengue_areas)

In dengue but NOT in weather:
{'Kalmune', 'Hambanthota', 'NuwaraEliya'}

In weather but NOT in dengue:
{'Hambantota', 'Kalmunai', 'Nuwara Eliya'}


In [35]:
print("In dengue but NOT in population:")
print(dengue_areas - population_areas)

print("\nIn population but NOT in dengue:")
print(population_areas - dengue_areas)

In dengue but NOT in population:
{'Kalmune', 'Hambanthota', 'NuwaraEliya'}

In population but NOT in dengue:
{'Hambantota', 'Nuwara-eliya'}


In [36]:
print("In weather but NOT in population:")
print(weather_areas - population_areas)

print("\nIn population but NOT in weather:")
print(population_areas - weather_areas)

In weather but NOT in population:
{'Kalmunai', 'Nuwara Eliya'}

In population but NOT in weather:
{'Nuwara-eliya'}


In [37]:
population_unique = population.drop_duplicates()

years_per_district = population_unique.groupby("district")["year"].nunique()

print(years_per_district)

district
Ampara          11
Anuradhapura    11
Badulla         11
Batticaloa      11
Colombo         11
Galle           11
Gampaha         11
Hambantota      11
Jaffna          11
Kalutara        11
Kandy           11
Kegalle         11
Kilinochchi     11
Kurunegala      11
Mannar          11
Matale          11
Matara          11
Monaragala      11
Mullaitivu      11
Nuwara-eliya    11
Polonnaruwa     11
Puttalam        11
Ratnapura       11
Trincomalee     11
Vavuniya        11
Name: year, dtype: int64


In [38]:
print("Minimum years per district:", years_per_district.min())
print("Maximum years per district:", years_per_district.max())

Minimum years per district: 11
Maximum years per district: 11


## Data Inventory Summary

In [39]:
summary = pd.DataFrame({
    "Dataset": ["Dengue", "Weather", "Population"],
    "Rows": [
        len(dengue),
        len(weather),
        len(population)
    ],
    "Columns": [
        len(dengue.columns),
        len(weather.columns),
        len(population.columns)
    ],
    "Start Year": [
        dengue["year"].min(),
        weather["year"].min(),
        population["year"].min()
    ],
    "End Year": [
        dengue["year"].max(),
        weather["year"].max(),
        population["year"].max()
    ],
    "Geographic Areas": [
        dengue["district"].nunique(),
        weather["district"].nunique(),
        population["district"].nunique()
    ],
    "Missing Values": [
        dengue.isnull().sum().sum(),
        weather.isnull().sum().sum(),
        population.isnull().sum().sum()
    ],
    "Duplicate Rows": [
        dengue.duplicated().sum(),
        weather.duplicated().sum(),
        population.duplicated().sum()
    ]
})

summary

,Dataset,Rows,Columns,Start Year,End Year,Geographic Areas,Missing Values,Duplicate Rows
0,Dengue,20306,6,2010,2024,26,0,0
1,Weather,20332,19,2010,2024,26,0,0
2,Population,475,3,2014,2024,25,0,200


## Key Findings from the Data Inventory

1. The dengue dataset contains weekly dengue case observations covering 2010–2024.

2. The weather dataset also covers 2010–2024 and contains temperature, humidity, rainfall and wind speed variables.

3. No standard null (`NaN`) values were detected in the three raw datasets. However, temporal completeness must be considered separately from cell level missing values.

4. Dengue Week 52 of 2023 is absent, while the corresponding week exists in the weather dataset.

5. The population dataset covers 2014–2024 rather than the complete 2010–2024 study period.

6. The population dataset contains 200 exact duplicate rows. After excluding exact duplicates for inspection, 275 unique records remain.

7. Geographic naming inconsistencies exist among the datasets, including variations of Hambantota and Nuwara Eliya.

8. The dengue/weather data contain an additional Kalmunai/Kalmune reporting area that requires investigation before integration with the 25 district population dataset.